pennylane/labs/lie/lie_closure.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A function to compute the Lie closure of a set of operators"""
+# pylint: disable=too-many-arguments
+
+import warnings
+from collections.abc import Iterable
+from typing import Union
+
+import numpy as np
+
+import pennylane as qml
+from pennylane.operation import Operator
+from pennylane.pauli import PauliSentence, PauliWord
+
+
+def _hermitian_basis(matrices, tol=None):
+    """Find a linear independent basis of a list of Hermitian matrices
+
+    Args:
+        matrices (Iterable[numpy.ndarray]): A list of Hermitian matrices.
+        tol (float): Tolerance for linear dependence check. Defaults to ``1e-10``.
+    """
+    if tol is None:
+        tol = 1e-10
+
+    basis = []
+    for A in matrices:
+        if not np.allclose(A.conj().T, A):
+            A = 1j * A
+            if not np.allclose(A.conj().T, A):
+                warnings.warn("Some basis matrices are not Hermitian", UserWarning)
+
+        B = A.copy()
+        for C in basis:
+            B -= np.trace(np.dot(C.conj().T, A)) * C
+        if np.linalg.norm(B) > tol:  # Tolerance for numerical stability
+            B /= np.linalg.norm(B)
+            basis.append(B)
+    return np.array(basis)
+
+
+def lie_closure_dense(
+    generators: Iterable[Union[PauliWord, PauliSentence, Operator]],
+    n=None,
+    max_iterations: int = 10000,
+    verbose: bool = False,
+    tol: float = None,
+):
+    r"""Compute the dynamical Lie algebra :math:`frak{g}` from a set of generators using their dense matrix representation.
+
+    This function computes the Lie closure of a set of generators using their dense matrix representation.
+    This is sometimes more efficient than using the sparse Pauli representations of :class:`~PauliWord` and
+    `~PauliSentence` that are employed in :func:`~lie_closure`, e.g., when ther are few but dense sums of Paulis.
+
+    .. seealso:: For details on the mathematical definitions, see :func:`~lie_closure` and our `Introduction to Dynamical Lie Algebras for quantum practitioners <https://pennylane.ai/qml/demos/tutorial_liealgebra/>`__.
+
+    Args:
+        generators (Iterable[Union[PauliWord, PauliSentence, Operator]]): generating set for which to compute the
+            Lie closure.
+        n (int): The number of qubits involved. If ``None`` is provided, it is automatically deduced from the generators.
```

Comment:
In order to make `qml.matrix` with `wire_order=range(n)` work, we need this, no?
```suggestion
        n (int): The number of qubits involved. If ``None`` is provided, it is automatically deduced from the generators.
            The involved qubits of all operators together should be integers from ``0`` to ``n-1``.
```

Reply:
Oh right a user could provide ops with non-integer valued ops 🙈 I like the suggestion to just not consider that atm (fair since this is experimental research code)